File: patterns/Behavioral/Command/JavaScript/RealWorldExample/WebScraping.js
class Command {
    execute() {}
    getId() {}
    getStatus() {}
}

class WebScrapingCommand extends Command {
    constructor(url) {
        super();
        this.id = null;
        this.status = 0;
        this.url = url;
    }

    getId() {
        return this.id;
    }

    getStatus() {
        return this.status;
    }

    getURL() {
        return this.url;
    }

    async execute() {
        const html = await this.download();
        this.parse(html);
        this.complete();
    }

    async download() {
        const response = await fetch(this.getURL());
        const html = await response.text();
        console.log(`WebScrapingCommand: Downloaded ${this.url}`);
        return html;
    }

    parse(html) {
        throw new Error("Method 'parse()' must be implemented.");
    }

    complete() {
        this.status = 1;
        Queue.get().completeCommand(this);
    }
}

class IMDBGenresScrapingCommand extends WebScrapingCommand {
    constructor() {
        super("https://www.imdb.com/feature/genre/");
    }

    parse(html) {
        const regex = /href="(https:\/\/www.imdb.com\/search\/title\?genres=.*?)"/g;
        const matches = [...html.matchAll(regex)];
        console.log(`IMDBGenresScrapingCommand: Discovered ${matches.length} genres.`);
        matches.forEach(match => {
            Queue.get().add(new IMDBGenrePageScrapingCommand(match[1]));
        });
    }
}

class IMDBGenrePageScrapingCommand extends WebScrapingCommand {
    constructor(url, page = 1) {
        super(url);
        this.page = page;
    }

    getURL() {
        return `${this.url}?page=${this.page}`;
    }

    parse(html) {
        const regex = /href="(\/title\/.*?\/)\?ref_=adv_li_tt"/g;
        const matches = [...html.matchAll(regex)];
        console.log(`IMDBGenrePageScrapingCommand: Discovered ${matches.length} movies.`);
        matches.forEach(match => {
            const url = `https://www.imdb.com${match[1]}`;
            Queue.get().add(new IMDBMovieScrapingCommand(url));
        });

        if (html.includes('Next &#187;</a>')) {
            Queue.get().add(new IMDBGenrePageScrapingCommand(this.url, this.page + 1));
        }
    }
}

class IMDBMovieScrapingCommand extends WebScrapingCommand {
    parse(html) {
        const regex = /<h1 itemprop="name" class="">(.*?)<\/h1>/;
        const match = html.match(regex);
        if (match) {
            const title = match[1];
            console.log(`IMDBMovieScrapingCommand: Parsed movie ${title}.`);
        }
    }
}

class Queue {
    constructor() {
        this.db = new Map();
        this.lastId = 0;
    }

    isEmpty() {
        return ![...this.db.values()].some(command => command.status === 0);
    }

    add(command) {
        this.lastId++;
        command.id = this.lastId;
        this.db.set(this.lastId, command);
    }

    getCommand() {
        for (let command of this.db.values()) {
            if (command.status === 0) {
                return command;
            }
        }
        return null;
    }

    completeCommand(command) {
        if (this.db.has(command.getId())) {
            this.db.get(command.getId()).status = command.getStatus();
        }
    }

    async work() {
        while (!this.isEmpty()) {
            const command = this.getCommand();
            if (command === null) {
                break;
            }
            await command.execute();
        }
    }

    static get() {
        if (!Queue.instance) {
            Queue.instance = new Queue();
        }
        return Queue.instance;
    }
}

// Client Code
const queue = Queue.get();
if (queue.isEmpty()) {
    queue.add(new IMDBGenresScrapingCommand());
}

queue.work();

File: patterns/Behavioral/Command/Go/RealWorldExample/WebScraping.go
package main

import (
	"database/sql"
	"encoding/base64"
	"encoding/gob"
	"fmt"
	"io/ioutil"
	"log"
	"regexp"

	_ "github.com/mattn/go-sqlite3"
)

// Command interface
type Command interface {
	Execute()
	GetID() int
	GetStatus() int
}

// WebScrapingCommand struct (Abstract)
type WebScrapingCommand struct {
	ID     int
	Status int
	URL    string
}

func (c *WebScrapingCommand) GetID() int {
	return c.ID
}

func (c *WebScrapingCommand) GetStatus() int {
	return c.Status
}

func (c *WebScrapingCommand) Download() string {
	resp, err := ioutil.ReadFile(c.URL) // Simulates file download
	if err != nil {
		log.Fatalf("Failed to download: %s", err)
	}
	fmt.Printf("WebScrapingCommand: Downloaded %s\n", c.URL)
	return string(resp)
}

func (c *WebScrapingCommand) Complete() {
	c.Status = 1
	QueueInstance.CompleteCommand(c)
}

// IMDBGenresScrapingCommand struct
type IMDBGenresScrapingCommand struct {
	WebScrapingCommand
}

func NewIMDBGenresScrapingCommand() *IMDBGenresScrapingCommand {
	return &IMDBGenresScrapingCommand{
		WebScrapingCommand{URL: "https://www.imdb.com/feature/genre/"},
	}
}

func (c *IMDBGenresScrapingCommand) Execute() {
	html := c.Download()
	c.Parse(html)
	c.Complete()
}

func (c *IMDBGenresScrapingCommand) Parse(html string) {
	re := regexp.MustCompile(`href="(https://www.imdb.com/search/title\?genres=.*?)"`)
	matches := re.FindAllStringSubmatch(html, -1)
	fmt.Printf("IMDBGenresScrapingCommand: Discovered %d genres.\n", len(matches))
	for _, match := range matches {
		QueueInstance.Add(&IMDBGenrePageScrapingCommand{
			WebScrapingCommand: WebScrapingCommand{URL: match[1]},
			Page:               1,
		})
	}
}

// IMDBGenrePageScrapingCommand struct
type IMDBGenrePageScrapingCommand struct {
	WebScrapingCommand
	Page int
}

func (c *IMDBGenrePageScrapingCommand) Execute() {
	html := c.Download()
	c.Parse(html)
	c.Complete()
}

func (c *IMDBGenrePageScrapingCommand) Parse(html string) {
	re := regexp.MustCompile(`href="(/title/.*?)/"`)
	matches := re.FindAllStringSubmatch(html, -1)
	fmt.Printf("IMDBGenrePageScrapingCommand: Discovered %d movies.\n", len(matches))
	for _, match := range matches {
		url := "https://www.imdb.com" + match[1]
		QueueInstance.Add(&IMDBMovieScrapingCommand{
			WebScrapingCommand: WebScrapingCommand{URL: url},
		})
	}
}

// IMDBMovieScrapingCommand struct
type IMDBMovieScrapingCommand struct {
	WebScrapingCommand
}

func (c *IMDBMovieScrapingCommand) Execute() {
	html := c.Download()
	c.Parse(html)
	c.Complete()
}

func (c *IMDBMovieScrapingCommand) Parse(html string) {
	re := regexp.MustCompile(`<h1 itemprop="name" class="">(.*?)</h1>`)
	match := re.FindStringSubmatch(html)
	if len(match) > 1 {
		title := match[1]
		fmt.Printf("IMDBMovieScrapingCommand: Parsed movie %s.\n", title)
	}
}

// Queue struct (Singleton)
type Queue struct {
	db *sql.DB
}

var QueueInstance *Queue

func init() {
	db, err := sql.Open("sqlite3", "./commands.db")
	if err != nil {
		log.Fatal(err)
	}

	QueueInstance = &Queue{db: db}
	QueueInstance.db.Exec(`CREATE TABLE IF NOT EXISTS commands (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		command BLOB,
		status INTEGER
	)`)
}

func (q *Queue) Add(command Command) {
	var buf []byte
	err := gob.NewEncoder(buf).Encode(command)
	if err != nil {
		log.Fatalf("Failed to serialize command: %s", err)
	}

	q.db.Exec("INSERT INTO commands (command, status) VALUES (?, ?)", base64.StdEncoding.EncodeToString(buf), command.GetStatus())
}

func (q *Queue) GetCommand() Command {
	row := q.db.QueryRow("SELECT * FROM commands WHERE status = 0 LIMIT 1")
	var id int
	var commandStr string
	var status int
	err := row.Scan(&id, &commandStr, &status)
	if err != nil {
		log.Fatal(err)
	}

	commandBytes, _ := base64.StdEncoding.DecodeString(commandStr)
	var command Command
	gob.NewDecoder(commandBytes).Decode(&command)
	return command
}

func (q *Queue) CompleteCommand(command Command) {
	q.db.Exec("UPDATE commands SET status = ? WHERE id = ?", command.GetStatus(), command.GetID())
}

func main() {
	QueueInstance.Add(NewIMDBGenresScrapingCommand())
	for {
		command := QueueInstance.GetCommand()
		if command == nil {
			break
		}
		command.Execute()
	}
}
